pyvista/plotting/plotting.py

```diff
@@ -1581,6 +1591,47 @@ def window_size(self):
     def window_size(self, window_size):
         """Set the render window size."""
         self.render_window.SetSize(window_size[0], window_size[1])
+        self.render()
+
+    @contextmanager
+    def window_size_context(self, window_size=None):
+        """Set the render window size in an isolated context.
+
+        Parameters
+        ----------
+        window_size : Sequence, optional
+            Window size in pixels.  Defaults to :attr:`pyvista.Plotter.window_size`.
+
+        Examples
+        --------
+        Take two different screenshots with two different window sizes.
+
+        >>> import pyvista as pv
+        >>> pl = pv.Plotter(off_screen=True)
+        >>> _ = pl.add_mesh(pv.Cube())
+        >>> with pl.window_size_context((400, 400)):
+        ...     pl.screenshot('/tmp/small_screenshot.png')  # doctest:+SKIP
+        >>> with pl.window_size_context((1000, 1000)):
+        ...     pl.screenshot('/tmp/big_screenshot.png')  # doctest:+SKIP
+
+        """
+        # No op if not set
+        if window_size is None:
+            yield self
+            return
+        # If render window is not current
+        if self.render_window is None:
+            warnings.warn('Attempting to set window_size on an unavailable render widow.')
+            yield self
+            return
+        size_before = self.window_size
+        if window_size is not None:
+            self.window_size = window_size
+        yield self
```

Comment:
IIUC note that (IIRC, unlike pytest) `@contextmanager` needs to use this pattern
```
try:
    yield
finally:
    set_whatever = back_to_original_value
```
https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

> At the point where the generator yields, the block nested in the [with](https://docs.python.org/3/reference/compound_stmts.html#with) statement is executed. The generator is then resumed after the block is exited. If an unhandled exception occurs in the block, it is reraised inside the generator at the point where the yield occurred. Thus, you can use a [try](https://docs.python.org/3/reference/compound_stmts.html#try)…[except](https://docs.python.org/3/reference/compound_stmts.html#except)…[finally](https://docs.python.org/3/reference/compound_stmts.html#finally) statement to trap the error (if any), or ensure that some cleanup takes place. If an exception is trapped merely in order to log it or to perform some action (rather than to suppress it entirely), the generator must reraise that exception. Otherwise the generator context manager will indicate to the with statement that the exception has been handled, and execution will resume with the statement immediately following the with statement.

Reply:
Okay woah. I confidently though that the `@contextmanager` handled this for us. I just tested this and indeed, encountering a exception indeed fails to set back the original value.

Sample code:

```py
from contextlib import contextmanager

class Contextual:
    def __init__(self):
        self.value = 1
        
    @contextmanager
    def context(self, value):
        before = self.value
        self.value = value
        yield
        self.value = before

c = Contextual()

with c.context(5):
    print(f'In context: {c.value}')
    raise RuntimeError()
```

Then after error:
```py
>>> c.value
5
```

I will open up a follow up PR to fix this

Thanks, @larsoner